### DISCUSSION TITLE
What am I doing wrong if my RasterizingLayer only redraws on pan/zoom?

### DISCUSSION THREAD

**Discussion by bergziege** [Mapsui/Mapsui#2338] [Q&A]
Hi,
using a `RasterizingLayer` to show some lines which change at runtime. To notify the layer about changes I call `DataHasChanged()` after updating the `DataSource` of the parent layer. Right after that the map goes blank until I pan or zoom the map. After paning/zooming everything draw just fine.

So what setting am I missing/do I have to call to trigger an "redraw" just after data changes?

App Info:
- WinAppSdk/WinUI3

**Comment by pauldendulk** [comment]
DataHasChanged() should trigger a Map.RefreshData(). Perhaps something goes wrong when propagating the event back to the map because the RasterizingLayer has parent and child layers.

Can you fix the problem by calling Map.RefreshData() instead of Layer.DataHasChanged()?

**Comment by bergziege** [comment]
Well, that solved part of the problem but introduced another. 

- It now loads on app start ... yeah
- But when updating the data I now have to klick in the map -> everything disappears -> pan view -> reappears
  - or skip the click and just pan. But its not refreshing without interaction.

I pushed a very reduced version of my app to [bergziege/MapsUiSandbox](https://github.com/bergziege/MapsUiSandbox) (sry for the chaos, just stripped everything away but the map).
The file in question is [MapViewModel](https://github.com/bergziege/MapsUiSandbox/blob/main/SegmentPlaner/Modules/Map/UI/MapViewModel.cs) around line 102 (and line 45 for the initialization).

To run the app you might have to change the path in App.config.
If the map sometimes does not show anything on startup ... thats a timing problem I am aware of.

**Reply by pauldendulk** [reply]
Thanks for providing the sample. I can run it and see a trail. So far so good. You say:

> But when updating the data I now have to klick in the map -> everything disappears -> pan view -> reappears

It is now clear to me right now when the data is updated, or how I could update the data when using the sample. But before you explain this to me perhaps you could first look at this issue which could be about the same problem:

https://github.com/Mapsui/Mapsui/issues/2345

If you modify data in v4 you need to use a layer that is derived from IModifyFeatureLayer and you need to clear the cache like IFeature.RenderedGeometry.Clear().

**Reply by bergziege** [reply]
Hi, thanks for having a look at it.

I checked the bug you mentioned but thats not really my use case (at least I cannot see it). So here goes some more details:
The final app will be a long distance cycle route planner based on preplanned segments. The user adds new segments in form of gpx files and then creates a new route by selecting the segments he wants to use.
The current use case is that the user adds new segments and wants to show them on the map. To do this click on "New", select another gpx file and then click on "Refresh" to update the map.
In this case no existing feature will be edited/updated but the layer will completely rebuild.

`MapViewModel` line 102:
```
GeometrySimplifyProvider simplifyProvider = new GeometrySimplifyProvider(new MemoryProvider(segmentFeatures));

segmentLayer.DataSource = simplifyProvider;
map.RefreshData();
```
where `segmentLayer` is a ordinary `Layer` object.

The above works perfectly (rebuilds the layer and shows the newly added segments) when using the `segmentLayer` directly (line 54):
```
map.Layers.Add(segmentLayer);
//map.Layers.Add(rasterizedSegmentLayer);
```

But as soon as I use the `rasterizedSegmentLayer` to gain some performance, _the map update will only happen when I pan/zoom the map after a click on "Refresh"_.

Maybe the problem is that I create new instances every time I update (assigning a new instance of `GeometrySimplifyProvider` to `segmentLayer.DataSource` - but do not know what else to do without everything stopping to work)?
The above problem does not occour on first app start where everything draws just fine once.
